File: src/llpm/llpm_pht2_bpred_adapter.cc
#include "llpm/llpm_pht2_bpred_adapter.hh"

#include "base/logging.hh"

namespace gem5
{

LLPMPHT2BPredAdapter::LLPMPHT2BPredAdapter(
    const LLPMPHT2BPredAdapterParams &params)
    : SimObject(params),
      component(params.component),
      libraryPath(params.library_path),
      resetCycles(params.reset_cycles),
      jsonConfig(params.json_config),
      entries(params.entries),
      initialCounter(params.initial_counter),
      resultsDir(params.results_dir)
{
    if (component != "rtl-pht2-bpred") {
        fatal("LLPMPHT2BPredAdapter only supports rtl-pht2-bpred");
    }
    if (initialCounter > 3) {
        fatal("LLPMPHT2BPredAdapter initial_counter must be a 2-bit value");
    }
}

} // namespace gem5

File: src/llpm/LLPMPHT2BPredAdapter.py
from m5.params import *
from m5.SimObject import SimObject


class LLPMPHT2BPredAdapter(SimObject):
    type = 'LLPMPHT2BPredAdapter'
    cxx_header = 'llpm/llpm_pht2_bpred_adapter.hh'
    cxx_class = 'gem5::LLPMPHT2BPredAdapter'

    component = Param.String('rtl-pht2-bpred', 'Selected LLPM predictor component')
    library_path = Param.String('', 'Path to the LLPM predictor shared library')
    reset_cycles = Param.Unsigned(2, 'Component reset cycles')
    json_config = Param.String('', 'JSON configuration passed to the LLPM component')
    entries = Param.Unsigned(1024, 'PHT entry count')
    initial_counter = Param.Unsigned(1, 'Initial 2-bit PHT counter value')
    results_dir = Param.String('', 'Directory for LLPM adapter result artifacts')
